pkg/dockerregistry/server/signaturedispatcher.go

```diff
@@ -0,0 +1,166 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+
+	ctxu "github.com/docker/distribution/context"
+
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/api/v2"
+	"github.com/docker/distribution/registry/handlers"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+
+	gorillahandlers "github.com/gorilla/handlers"
+)
+
+type signature struct {
+	// Name must be in "sha256:<digest>@signatureName" format
+	Name string `json:"name"`
+	// Type is optional, of not set it will be defaulted to "AtomicImageV1"
+	Type string `json:"type"`
+	// Content contains the signature
+	Content []byte `json:"content"`
+}
+
+type signatureList struct {
+	Signatures []signature `json:"signatures"`
+}
+
+const errGroup = "registry.api.v2"
+
+var (
+	ErrorCodeSignatureInvalid = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "SIGNATURE_INVALID",
+		Message:        "invalid image signature",
+		HTTPStatusCode: http.StatusBadRequest,
+	})
+
+	ErrorCodeSignatureAlreadyExists = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "SIGNATURE_EXISTS",
+		Message:        "image signature already exists",
+		HTTPStatusCode: http.StatusConflict,
+	})
+)
+
+type signatureHandler struct {
+	ctx       *handlers.Context
+	reference imageapi.DockerImageReference
+}
+
+// SignatureDispatcher handles the GET and PUT requests for signature endpoint.
+func SignatureDispatcher(ctx *handlers.Context, r *http.Request) http.Handler {
+	signatureHandler := &signatureHandler{ctx: ctx}
+	signatureHandler.reference, _ = imageapi.ParseDockerImageReference(ctxu.GetStringValue(ctx, "vars.name") + "@" + ctxu.GetStringValue(ctx, "vars.digest"))
```

Comment:
If the parsing fails, should this become an error handler?  Or should you check reference empty in signatureHandler and error?